src/watchdog.c

```diff
@@ -129,6 +135,8 @@ IMPLEMENT_MOCKABLE_FUNCTION(, void, watchdog_reset, WATCHDOG_HANDLE, watchdog)
     }
     else
     {
+        /* Codes_SRS_WATCHDOG_45_003: [ watchdog_reset shall wait until state is not EXPIRING. ]*/
+        (void)InterlockedHL_WaitForNotValue(&watchdog->state, WATCHDOG_EXPIRING, UINT32_MAX);
         /*Codes_SRS_WATCHDOG_42_032: [ watchdog_reset shall set the state of the watchdog to STOP. ]*/
         (void)interlocked_exchange(&watchdog->state, WATCHDOG_STOP);

```

Comment:
This is racy. 
Imagine a watchdog with state RUNNING.
thread T1 getting stopped between ln 139 and 141 (it passed the check that state is not EXPIRING.

Thread T2 executes the callback which changes the state to EXPIRING. Right after changing the state it gets switched out to T1.

T1 continues executing, setting the state to STOP, freeing stuff.

T2 resumes execution and winamps.

Reply:
I think this would have to be done with interlocked_compare_exchange

Reply:
The comment about the fact that we'd free memory is applicable to watchdog_stop.
But the state would be inconsistent also for watchdog_reset, just in different ways.

Reply:
I'm expecting SM to handle this for me.